### ISSUE TITLE
Use module_method for RNNCellBase.initialize_carry

### ISSUE THREAD

**Issue by ronw** [google/flax#91]
The RNNCellBase interface requires passing consistent arguments to both initialize_carry and apply.   if the caller isn't careful to pass the same args (e.g. cell_size) to both methods, things will immediately break.  If these hyperparams were instead specified only once, users couldn't make that mistake.

(This isn't an issue for the existing cells because they infer cell sizes from the carry argument, but this issue can come up for more complex cells.)

If initialize_carry was defined as a module_method, shared hyperparameters could be applied using .partial()

This would also make it possible for an RNN to learn its initial state.

**Comment by david-waterworth** [comment]
I think this is a good idea. I tend to provide my own initialize_carry function, this could be one of

- Constant
- Learnable
- Another model which attempts to predict x0 from previous inputs and outputs

**Comment by marcvanzee** [comment]
@bastings has done some experiments on this issue and they had some insights that I think would be useful to share here. I've assigned this issue to them.

**Comment by levskaya** [comment]
Closing this issue as the old `nn` library is deprecated and we hopefully have a better pattern here in the newer "linen" api:

Just define a non-static `initialize_carry`  e.g. `def initialize_carry(self, rng, batch_dims, size, init_fn=zeros):` in linen and it will have access to the Module constructor args on `self`.  

Please feel free to re-open (or better make a new issue) pertaining to any perceived deficiency in linen's RNN utils.
